docs/src/conventions.md

```diff
@@ -101,8 +101,51 @@ based on the following rules.
   (or `wrap_array_native(u_ode, semi)`) for further processing.
 - When some solution is passed together with the `mesh, equations, solver, cache, ...`,
   it is already wrapped via `wrap_array` (or `wrap_array_native`).
-- Exceptions of this rule are possible, e.g. for AMR, but must be documented in
+- Exceptions of this rule are possible, e.g., for AMR, but must be documented in
   the code.
 - `wrap_array` should be used as default option. `wrap_array_native` should only
   be used when necessary, e.g., to avoid additional overhead when interfacing
   with external C libraries such as HDF5, MPI, or visualization.
+
+## Numeric types and type stability
+In Trixi.jl, `Float32` and `Float64` types are fully supported. We ensure the type stability of these numeric types throughout the development process. Below are some guidelines to apply in various scenarios.
+
+- **Exact floating-point numbers**: Some real numbers can be represented exactly in machine (e.g., `0.25`, `0.5`, `1/2`) and we prefer to use `Float32` type of them to achieve a concise way of possible type promotion. For example,
+  ```julia
+  # Assume we have `0.25`, `0.5`, `1/2` in function
+  0.25f0, 0.5f0, 0.5f0 # corresponding numbers
+  ```
+
+- **Non-exact floating-point numbers**: For real numbers that cannot be exactly represented in machine precision (e.g., `0.1`, `1/3`, `pi`), use the `convert` function to make them consistent with the type of the function input. For example, 
+  ```julia
+  # Assume we are handling `pi` in function
+  function foo(input1, input2, input3, ...)
+    RealT = eletype(input1) # good practice to always use the first input to extract numeric type
+    # ...
+    c1 = convert(RealT, pi) * c2 # sample operation
+    # ...
```

Comment:
```suggestion
  function foo(u, input2, input3, ...)
    RealT = eltype(u) # good practice to always use the first input to extract numeric type
    # ...
    c1 = convert(RealT, pi) * c2 # sample operation
    # ...
```

Maybe like this to clarify that we typically want to use the same number type as the unknowns `u`?

Reply:
What is the difference? The original one suggests a way to better use the first input, as otherwise people might think we should use an explicit `u` to get the number types. This could be confusing since some functions do not take `u` as input.

Reply:
Fair enough. What about being even more explicit here since not all functions have a first argument where it would make sense to get a real type from? Concretely, we could basically leave your suggestion with the fix

```suggestion
  function foo(input1, input2, input3, ...)
    RealT = eltype(input1) # good practice to always use the first input to extract numeric type
    # ...
    c1 = convert(RealT, pi) * c2 # sample operation
    # ...
```

and list some recommendations later:

1. If the function gets a local pointwise vector of the solution variables `u` such as `flux(u, equations)`, use `u` to determine the real type `eltype(u)`.
2. If `u` is not passed as an argument but a vector of coordinates `x` such as `initial_condition(x, t, equations)`, use `eltype(x)` instead.
3. Choose an appropriate argument to determine the real type otherwise.